### PR TITLE
Add a basic unit test for relay_node

### PR DESCRIPTION
**Pull request by mllofriu**
I'm trying to implement the feature request I created here - https://github.com/ros-tooling/topic_tools/issues/58.

In doing so, I found myself refactoring code in tool_base_node.

In order to prevent breaking things, I figured I'd write some unit tests for the existing functionality. 

I'm creating this pull request as a draft to get some early feedback. If we agree on it, I can apply the same pattern to the other tools in topic_tools.

### CODE REVIEW

File: topic_tools/test/test_relay.cpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  auto subscription = test_node->create_subscription<std_msgs::msg::String>(
+    "/relay_test/output", 10, std::bind(&TestRelay::topic_callback, this, _1));
+  auto publisher = test_node->create_publisher<std_msgs::msg::String>("/relay_test/input", 10);
+  rclcpp::spin_some(test_node);
+
+  auto options = rclcpp::NodeOptions{};
+  options.append_parameter_override("input_topic", "/relay_test/input");
+  options.append_parameter_override("output_topic", "/relay_test/output");
+  auto target_node = std::make_shared<topic_tools::RelayNode>(options);
+  rclcpp::spin_some(target_node);
```
Comment (christophebedard) [inline topic_tools/test/test_relay.cpp:54]: why do you need to call `spin_some()` after creating the nodes?

File: topic_tools/test/test_relay.cpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  for (std::string msg_content : {"hello", "again"}) {
+    expected_msg = msg_content;
+
+    auto message = std_msgs::msg::String();
+    message.data = msg_content;
+    publisher->publish(message);
+
+    rclcpp::spin_some(test_node);
+    rclcpp::spin_some(target_node);
+    rclcpp::spin_some(test_node);
```
Comment (christophebedard) [inline topic_tools/test/test_relay.cpp:65]: Just wondering: why do you need to call `spin_some()` with the subscription node twice? And why do you need to call `spin_some()` for the publisher node after publishing?

File: topic_tools/CMakeLists.txt [hunk 3]
```diff
@@ -87,6 +87,20 @@ target_link_libraries(mux
...
+
+  find_package(ament_cmake_gtest REQUIRED)
+  ament_add_gtest(test_relay test/test_relay.cpp)
+  target_include_directories(test_relay PUBLIC
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+    $<INSTALL_INTERFACE:include>
+  )
+  ament_target_dependencies(test_relay
+    rclcpp
+    std_msgs
```
Comment (christophebedard) [inline topic_tools/CMakeLists.txt:99]: you'll need to add `find_package(std_msgs REQUIRED) above and add `std_msgs` it as a `<test_depend>` in the `package.xml` file

**Review by christophebedard — CHANGES_REQUESTED** [review]

**Comment by christophebedard** [comment]
Also, can you sign-off your commit? See https://github.com/ros-tooling/topic_tools/pull/59/checks?check_run_id=12844845270

File: topic_tools/test/test_relay.cpp (hunk 1 above)
Comment (mllofriu) [reply topic_tools/test/test_relay.cpp:54]: They were indeed unnecessary. I was following the logic that a listener needs a spin to find the topic, but that's clearly wrong. I need to read more on how this works under the hood.

File: topic_tools/CMakeLists.txt (hunk 3 above)
Comment (mllofriu) [reply topic_tools/CMakeLists.txt:99]: Thanks! Done.

**Comment by mllofriu** [comment]
Thanks a lot for the feedback. I'll push my changes to get a new build (will figure out how to trigger it).

Afterwards, I'll decline this PR and open one with tests for the other nodes in this package. Do you agree @christophebedard?

File: topic_tools/test/test_relay.cpp (hunk 1 above)
Comment (mllofriu) [reply topic_tools/test/test_relay.cpp]: Done.

File: topic_tools/test/test_relay.cpp (hunk 2 above)
Comment (mllofriu) [reply topic_tools/test/test_relay.cpp]: Indeed, the first call to `rclcpp::spin_some(test_node);` was not needed. This is related to my response to the other comment above. I ended up needing one `rclcpp::spin_some(target_node);` to trigger the callback, and then one `rclcpp::spin_some(test_node);` for the same reason. See commit 55b9ff2d693ab.

File: topic_tools/package.xml [hunk 4]
```diff
@@ -26,6 +26,8 @@
 
   <test_depend>ament_lint_auto</test_depend>
   <test_depend>ament_lint_common</test_depend>
+  <test_depend>ament_cmake_gtest</test_depend>
```
Comment (christophebedard) [inline topic_tools/package.xml:29]: it's a bit nitpicky, but you should order the entries here

```suggestion
  <test_depend>ament_cmake_gtest</test_depend>
  <test_depend>ament_lint_auto</test_depend>
  <test_depend>ament_lint_common</test_depend>
```

**Review by christophebedard — CHANGES_REQUESTED** [review]

**Comment by christophebedard** [comment]
> Afterwards, I'll decline this PR and open one with tests for the other nodes in this package. Do you agree @christophebedard?

Sure, that's fine!

**Pull request closed without merging**